File: BPMOPMobile_VietNamAirLines/BPMOPMobile_VietNamAirLines/BPMOPMobile.Droid/Presenter/SharedView/SharedView_PopupControlSingleLookup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.Graphics;
using Android.Graphics.Drawables;
using Android.OS;
using Android.Runtime;
using Android.Support.V7.App;
using Android.Support.V7.Widget;
using Android.Views;
using Android.Widget;
using BPMOPMobile.Bean;
using BPMOPMobile.Droid.Class;
using BPMOPMobile.Droid.Core.Common;
using BPMOPMobile.Droid.Presenter.Adapter;
using BPMOPMobile.Droid.Presenter.Fragment;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace BPMOPMobile.Droid.Presenter.SharedView
{
    public class SharedView_PopupControlSingleLookup : SharedView_PopupControlBase
    {
        public SharedView_PopupControlSingleLookup(LayoutInflater _inflater, AppCompatActivity _mainAct, CustomBaseFragment _fragment, string _fragmentTag, View _rootView) : base(_inflater, _mainAct, _fragment, _fragmentTag, _rootView)
        {

        }

        public override void InitializeValue_Master(ViewElement clickedElement)
        {
            base.InitializeValue_Master(clickedElement);
        }

        public override void InitializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JObject JObjectChild)
        {
            base.InitializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
        }

        public override void InitializeView()
        {
            base.InitializeView();
            try
            {
                #region Get View - Init Data

                View _viewPopupControl = _inflater.Inflate(Resource.Layout.PopupControl_SingleChoice, null);
                ImageView _imgClose = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_SingleChoice_Close);
                ImageView _imgDelete = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_SingleChoice_Delete);
                TextView _tvTitle = _viewPopupControl.FindViewById<TextView>(Resource.Id.tv_PopupControl_SingleChoice_Title);
                RecyclerView _recyData = _viewPopupControl.FindViewById<RecyclerView>(Resource.Id.recy_PopupControl_SingleChoice_Data);
                ImageView _imgDone = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_SingleChoice_Done);
                LinearLayout _lnSearch = _viewPopupControl.FindViewById<LinearLayout>(Resource.Id.ln_PopupControl_SingleChoice_Search);
                EditText _edtSearch = _viewPopupControl.FindViewById<EditText>(Resource.Id.edt_PopupControl_SingleChoice_Search);

                _imgDone.Visibility = ViewStates.Invisible;
                _imgDone.LayoutParameters.Width = 0;

                _lnSearch.Visibility = ViewStates.Gone;

                if (elementParent.Enable)
                {
                    _imgDelete.Visibility = ViewStates.Visible;
                }
                else
                {
                    _imgDelete.Visibility = ViewStates.Gone;
                }

                List<BeanLookupData> _lstLookupData = new List<BeanLookupData>();

                switch (base.flagView)
                {
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                    default:
                        {
                            _tvTitle.Text = elementParent.Title;

                            if (!String.IsNullOrEmpty(elementParent.DataSource))
                            {
                                _lstLookupData = JsonConvert.DeserializeObject<List<BeanLookupData>>(elementParent.DataSource);
                                if (_lstLookupData != null && _lstLookupData.Count > 0 && !String.IsNullOrEmpty(elementParent.Value)) // Value có giá trị mới làm
                                {
                                    for (int i = 0; i < _lstLookupData.Count; i++)
                                    {
                                        if (elementParent.Value.Contains(_lstLookupData[i].ID))
                                        {
                                            _lstLookupData[i].IsSelected = true;
                                            break; // vì chỉ có 1 thằng dc chọn nên break luôn
                                        }
                                    }
                                }
                            }
                            break;
                        }
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                        {
                            _tvTitle.Text = elementPopup.Title;

                            if (!String.IsNullOrEmpty(elementPopup.DataSource))
                            {
                                _lstLookupData = JsonConvert.DeserializeObject<List<BeanLookupData>>(elementParent.DataSource);
                                if (_lstLookupData != null && _lstLookupData.Count > 0 && !String.IsNullOrEmpty(elementParent.Value)) // Value có giá trị mới làm
                                {
                                    for (int i = 0; i < _lstLookupData.Count; i++)
                                    {
                                        if (elementParent.Value.Contains(_lstLookupData[i].ID))
                                        {
                                            _lstLookupData[i].IsSelected = true;
                                            break; // vì chỉ có 1 thằng dc chọn nên break luôn
                                        }
                                    }
                                }
                            }
                            break;
                        }
                }

                #endregion

                #region Show View                
                Dialog _dialogPopupControl = new Dialog(_rootView.Context, Resource.Style.Theme_Custom_BPMOP_Dialog_FullScreen);
                Window window = _dialogPopupControl.Window;
                _dialogPopupControl.RequestWindowFeature(1);
                _dialogPopupControl.SetCanceledOnTouchOutside(false);
                _dialogPopupControl.SetCancelable(true);
                window.SetSoftInputMode(SoftInput.AdjustPan); // Để Hide soft keyboard ko bị Resize
                window.SetGravity(GravityFlags.Bottom);
                var dm = _mainAct.Resources.DisplayMetrics;

                _dialogPopupControl.SetContentView(_viewPopupControl);
                _dialogPopupControl.Show();
                WindowManagerLayoutParams s = window.Attributes;
                s.Width = WindowManagerLayoutParams.MatchParent;
                s.Height = WindowManagerLayoutParams.MatchParent;
                window.Attributes = s;
                window.SetBackgroundDrawable(new ColorDrawable(Color.Transparent));
                #endregion

                #region Event
                _imgClose.Click += (sender, e) =>
                {
                    _dialogPopupControl.Dismiss();
                };

                _imgDelete.Click += (sender, e) =>
                {
                    switch (base.flagView)
                    {
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                        default:
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForElement(elementParent, "");
                                _frag._adapterDetailExpandControl.NotifyDataSetChanged();
                                _dialogPopupControl.Dismiss();
                            }
                            break;
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForPopupGridDetail(elementParent, elementPopup, JObjectChild, "");
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                    }
                };

                _edtSearch.TextChanged += (sender, e) =>
                {
                    List<BeanLookupData> _lstSearch = new List<BeanLookupData>();
                    if (!String.IsNullOrEmpty(_edtSearch.Text))
                    {
                        _lstSearch = _lstLookupData.Where(x => x.Title.ToLowerInvariant().Contains(_edtSearch.Text.ToLowerInvariant())).ToList();
                    }
                    else // Full
                    {
                        _lstSearch = _lstLookupData.ToList();
                    }
                    AdapterFormControlSingleChoice _adapterFormControlSingleChoice = new AdapterFormControlSingleChoice((MainActivity)base._mainAct, _rootView.Context, _lstSearch);
                    StaggeredGridLayoutManager staggeredGridLayoutManager = new StaggeredGridLayoutManager(1, LinearLayoutManager.Vertical);
                    _recyData.SetAdapter(_adapterFormControlSingleChoice);
                    _recyData.SetLayoutManager(staggeredGridLayoutManager);

                    if (elementParent.Enable)
                    {
                        _adapterFormControlSingleChoice.CustomItemClick += (sender, e) =>
                        {
                            BeanLookupData _selectedLookupItem = e;
                            List<BeanLookupData> _lstSelected = new List<BeanLookupData>();
                            _lstSelected.Add(_selectedLookupItem);

                            if (!String.IsNullOrEmpty(_selectedLookupItem.Title))
                            {
                                switch (base.flagView)
                                {
                                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                                    default:
                                        {
                                            FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                            _frag.UpdateValueForElement(elementParent, JsonConvert.SerializeObject(_lstSelected));
                                            _frag._adapterDetailExpandControl.NotifyDataSetChanged();
                                            break;
                                        }
                                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                                        {
                                            FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                            _frag.UpdateValueForPopupGridDetail(elementParent, elementPopup, JObjectChild, JsonConvert.SerializeObject(_lstSelected));
                                            break;
                                        }
                                }
                            }
                            _dialogPopupControl.Dismiss();
                        };
                    }
                };
                #endregion

                _edtSearch.Text = "";
            }
            catch (Exception ex)
            {
#if DEBUG
                CmmDroidFunction.WriteTrackingError(this.GetType().Name, "InitializeView", ex);
#endif
            }
        }
    }
}

File: VNABPMOP/app/src/main/java/com/vuthao/bpmop/shareview/SharedView_PopupControlSingleLookup.java
package com.vuthao.bpmop.shareview;

import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import com.google.gson.Gson;
import com.vuthao.bpmop.R;
import com.vuthao.bpmop.base.BroadcastUtility;
import com.vuthao.bpmop.base.Functions;
import com.vuthao.bpmop.base.activity.BaseActivity;
import com.vuthao.bpmop.base.api.ApiController;
import com.vuthao.bpmop.base.model.ApiList;
import com.vuthao.bpmop.base.model.custom.LookupData;
import com.vuthao.bpmop.base.model.dynamic.ViewElement;
import com.vuthao.bpmop.base.vars.VarsReceiver;
import com.vuthao.bpmop.core.Vars;
import com.vuthao.bpmop.shareview.adapter.FormControlSingleChoiceAdapter;

import org.json.JSONObject;

import java.util.ArrayList;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SharedView_PopupControlSingleLookup extends SharedView_PopupControlBase implements TextWatcher, View.OnClickListener {
    private ArrayList<LookupData> lstLookupData = new ArrayList<>();
    private FormControlSingleChoiceAdapter adapter;
    private Dialog dialog;

    private View view;
    private LinearLayout lnProgressBar;
    private LinearLayout lnSearch;
    private ImageView imgClose;
    private ImageView imgDelete;
    private TextView tvTitle;
    private RecyclerView recyData;
    private ImageView imgDone;
    private EditText edtSearch;
    private TextView tvNodata;

    public SharedView_PopupControlSingleLookup(LayoutInflater inflater, Activity mainAct, String fragmentTag, View rootView) {
        super(inflater, mainAct, fragmentTag, rootView);
    }

    @Override
    public void initializeValue_Master(ViewElement viewElement) {
        super.initializeValue_Master(viewElement);
    }

    @Override
    public void initializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JSONObject JObjectChild) {
        super.initializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
    }

    @Override
    public void initializeView() {
        super.initializeView();

        view = inflater.inflate(R.layout.popup_control_single_choice, null);
        lnProgressBar = view.findViewById(R.id.lnProgressbar);
        lnSearch = view.findViewById(R.id.ln_PopupControl_SingleChoice_Search);
        imgClose = view.findViewById(R.id.img_PopupControl_SingleChoice_Close);
        imgDelete = view.findViewById(R.id.img_PopupControl_SingleChoice_Delete);
        tvTitle = view.findViewById(R.id.tv_PopupControl_SingleChoice_Title);
        recyData = view.findViewById(R.id.recy_PopupControl_SingleChoice_Data);
        imgDone = view.findViewById(R.id.img_PopupControl_SingleChoice_Done);
        edtSearch = view.findViewById(R.id.edt_PopupControl_SingleChoice_Search);
        tvNodata = view.findViewById(R.id.tv_PopupControl_SingleChoice_NoData);

        edtSearch.setHint(Functions.share.getTitle("TEXT_SEARCH", "Search..."));

        lnSearch.setVisibility(View.VISIBLE);
        lnProgressBar.setVisibility(View.VISIBLE);
        imgDone.setVisibility(View.INVISIBLE);
        imgDone.getLayoutParams().width = 0;

        if (elementParent.isEnable()) {
            imgDelete.setVisibility(View.VISIBLE);
        } else {
            imgDelete.setVisibility(View.GONE);
        }

        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow:
            default: {
                tvTitle.setText(elementParent.getTitle());
                getDataSource(elementParent);
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                tvTitle.setText(elementPopup.getTitle());
                getDataSource(elementPopup);
                break;
            }
        }

        adapter = new FormControlSingleChoiceAdapter(mainAct, mainAct.getApplicationContext(), lstLookupData, pos -> {
            if (elementParent.isEnable()) {
                LookupData selectedLookupItem = lstLookupData.get(pos);
                ArrayList<LookupData> lstSelected = new ArrayList<>();
                lstSelected.add(selectedLookupItem);

                switch (flagView) {
                    case Vars.FlagViewControlDynamic.DetailWorkflow:
                    default: {
                        Intent intent = new Intent();
                        intent.setAction(VarsReceiver.UPDATEFORM);
                        intent.putExtra("element", new Gson().toJson(elementParent));
                        intent.putExtra("newValue", new Gson().toJson(lstSelected));
                        BroadcastUtility.send(mainAct, intent);
                        break;
                    }
                    case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                        Intent intent = new Intent();
                        intent.setAction(VarsReceiver.UPDATECHILDACTION);
                        intent.putExtra("elementParent", new Gson().toJson(elementParent));
                        intent.putExtra("elementChild", new Gson().toJson(elementPopup));
                        intent.putExtra("json", JObjectChild.toString());
                        intent.putExtra("newValue", new Gson().toJson(lstSelected));
                        BroadcastUtility.send(mainAct, intent);
                        break;
                    }
                }
            }

            dialog.dismiss();
        });

        StaggeredGridLayoutManager staggeredGridLayoutManager = new StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.VERTICAL);
        recyData.setAdapter(adapter);
        recyData.setLayoutManager(staggeredGridLayoutManager);

        // region Dialog
        dialog = new Dialog(mainAct, R.style.Theme_Custom_BPMOP_Dialog_FullScreen);
        Window window = dialog.getWindow();
        dialog.requestWindowFeature(1);
        dialog.setCanceledOnTouchOutside(false);
        dialog.setCancelable(true);
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
        window.setGravity(Gravity.BOTTOM);
        dialog.setContentView(view);
        dialog.show();

        WindowManager.LayoutParams s = window.getAttributes();
        s.width = WindowManager.LayoutParams.MATCH_PARENT;
        s.height = WindowManager.LayoutParams.MATCH_PARENT;
        window.setAttributes(s);
        // endregion

        imgClose.setOnClickListener(this);
        imgDelete.setOnClickListener(this);
        edtSearch.addTextChangedListener(this);
    }

    private void getDataSource(ViewElement element) {
        if (element.getListProprety() != null && element.getListProprety().size() > 0) {
            String lookupWebId = element.getListProprety().get(0).values().toArray()[0].toString();
            String lookupList = element.getListProprety().get(1).values().toArray()[0].toString();
            String lookupField = element.getListProprety().get(2).values().toArray()[0].toString();
            Call<ApiList<LookupData>> call = new ApiController().getApiRouteToken(BaseActivity.getToken()).getDataSource(lookupWebId, lookupList, lookupField);
            call.enqueue(new Callback<ApiList<LookupData>>() {
                @Override
                public void onResponse(@NonNull Call<ApiList<LookupData>> call, @NonNull Response<ApiList<LookupData>> response) {
                    assert response.body() != null;
                    lstLookupData = new ArrayList<>(response.body().getData());
                    switch (flagView) {
                        default:
                        case Vars.FlagViewControlDynamic.DetailWorkflow: {
                            for (int i = 0; i < lstLookupData.size(); i++) {
                                if (elementParent.getValue().contains(new Gson().toJson(lstLookupData.get(i)))) {
                                    lstLookupData.get(i).setSelected(true);
                                    break;
                                }
                            }
                            break;
                        }
                        case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                            for (int i = 0; i < lstLookupData.size(); i++) {
                                if (elementPopup.getValue().contains(new Gson().toJson(lstLookupData.get(i)))) {
                                    lstLookupData.get(i).setSelected(true);
                                    break;
                                }
                            }
                            break;
                        }
                    }

                    adapter.setList(lstLookupData);
                    edtSearch.setText("");
                    lnProgressBar.setVisibility(View.GONE);
                }

                @Override
                public void onFailure(@NonNull Call<ApiList<LookupData>> call, @NonNull Throwable t) {
                    lnProgressBar.setVisibility(View.GONE);
                }
            });
        } else {
            lnSearch.setVisibility(View.GONE);
            lnProgressBar.setVisibility(View.GONE);
            tvNodata.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void afterTextChanged(Editable s) {
        if (adapter != null) {
            adapter.getFilter().filter(s.toString());
        }
    }

    private void delete() {
        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow:
            default: {
                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATEFORM);
                intent.putExtra("element", new Gson().toJson(elementParent));
                intent.putExtra("newValue", "");
                BroadcastUtility.send(mainAct, intent);
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATECHILDACTION);
                intent.putExtra("elementParent", new Gson().toJson(elementParent));
                intent.putExtra("elementChild", new Gson().toJson(elementPopup));
                intent.putExtra("json", JObjectChild.toString());
                intent.putExtra("newValue", "");
                BroadcastUtility.send(mainAct, intent);
                break;
            }
        }

        dialog.dismiss();
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.img_PopupControl_SingleChoice_Close: {
                dialog.dismiss();
                break;
            }
            case R.id.img_PopupControl_SingleChoice_Delete: {
                delete();
                break;
            }
        }
    }
}
